torch/export/_draft_export.py

```diff
@@ -0,0 +1,177 @@
+    When exporting, we were unable to figure out if the expression `{self.data["expr"]}` always holds.
+    This occurred on the following line: {self.data["stack"]}.
+    As a result, it was specialized to evaluate to `{self.data["result"]}`, and asserts were inserted into the graph.
+
+    Please add `torch._check(...)` to the original code to assert this data-dependent assumption.
+    Please refer to https://docs.google.com/document/d/1kZ_BbB3JnoLbUZleDT6635dHs88ZVYId8jT-yTFgf3A/edit#heading=h.boi2xurpqa0o for more details.
+"""
+
+        else:
+            raise ValueError(f"Unknown failure type: {self.failure_type}")
+
+
+class DraftExportReport:
+    def __init__(self, failures: List[FailureReport]):
+        self.failures: List[FailureReport] = failures
+
+    def __str__(self):
+        if len(self.failures) == 0:
+            return """
+##############################################################################################
+Congratuations: No issues are found during export, and it was able to soundly produce a graph.
+You can now change back to torch.export.export()
+##############################################################################################
+"""
+
+        error = f"""
+###################################################################################################
+WARNING: {len(self.failures)} issue(s) found during export, and it was not able to soundly produce a graph.
+Please follow the instructions to fix the errors.
+Issues are compiled in hive table: TODO
+###################################################################################################
+
+"""
+
+        for i, failure in enumerate(self.failures):
+            error += f"{i + 1}. {str(failure)}\n"
+
+        return error
+
+    def serialize(self):
+        pass
+
+    def apply_suggested_fixes(self):
+        pass
+
+
+def draft_export(
+    mod: torch.nn.Module,
+    args: Tuple[Any, ...],
+    kwargs: Optional[Dict[str, Any]] = None,
+    *,
+    dynamic_shapes: Optional[Union[Dict[str, Any], Tuple[Any], List[Any]]] = None,
+    preserve_module_call_signature: Tuple[str, ...] = (),
+    fake_tensor_propagate_real_tensors: Optional[bool] = True,
+) -> ExportedProgram:
+    kwargs = kwargs or {}
+
+    with OperatorProfilingMode() as mode:
+        mod(*args, **kwargs)
+    custom_op_profile = mode.reports()
+
+    torch._dynamo.config.custom_ops_profile = custom_op_profile
+
+    with torch._functorch.config.patch(
+        fake_tensor_propagate_real_tensors=fake_tensor_propagate_real_tensors
+    ), RealTensorLoggingMode():
+        try:
+            new_shapes = None
+            ep = torch.export.export(
+                mod,
+                args,
+                kwargs,
+                dynamic_shapes=dynamic_shapes,
+                strict=False,
+                preserve_module_call_signature=preserve_module_call_signature,
+            )
+        except torch._dynamo.exc.UserError as exc:
```

Comment:
this looks like the "unsound" model follows the specs inferred from the model code, not the user specs